portal-bridge/src/types/state/trie_walker.rs

```diff
@@ -0,0 +1,155 @@
+}
+
+/// This struct takes in a root hash and a hashmap of changed nodes, then you can call an iterator
+/// which will return every proof to gossip
+pub struct TrieWalker {
+    pub nodes: BrownHashMap<B256, Vec<u8>>,
+    paths_to_parent: BrownHashMap<B256, PathToParent>,
+}
+
+impl TrieWalker {
+    pub fn new(root_hash: B256, nodes: BrownHashMap<B256, Vec<u8>>) -> Self {
+        let paths_to_parent = Self::get_paths_backwards(root_hash, &nodes);
+        Self {
+            nodes,
+            paths_to_parent,
+        }
+    }
+
+    fn get_paths_backwards(
+        root_hash: B256,
+        nodes: &BrownHashMap<B256, Vec<u8>>,
+    ) -> BrownHashMap<B256, PathToParent> {
+        let mut paths_to_parent: BrownHashMap<B256, PathToParent> = BrownHashMap::new();
+        let mut stack = vec![root_hash];
+        while let Some(node_key) = stack.pop() {
+            let encoded_node = nodes
+                .get(&node_key)
+                .expect("The stack should only contain nodes that are in the changed nodes");
+
+            let decoded_node = decode_node(&mut encoded_node.as_slice())
+                .expect("Should should only be passing valid encoded nodes");
+
+            match decoded_node {
+                Node::Extension(extension) => {
+                    let extension = extension.read().expect("Reading an extension should work");
+                    if let Node::Hash(hash_node) = &extension.node {
+                        // We could be walking a partial trie, so we need to check if the node is in
+                        // the changed nodes, if it is process it
+                        if nodes.contains_key(&hash_node.hash) {
+                            stack.push(hash_node.hash);
+                            paths_to_parent.insert(
+                                hash_node.hash,
+                                PathToParent::new(node_key, extension.prefix.get_data().to_vec()),
+                            );
+                        }
+                    }
+                }
+                Node::Branch(branch) => {
+                    let branch = branch.read().expect("Reading a branch should work");
+                    for (i, child) in branch.children.iter().enumerate() {
+                        if let Node::Hash(hash_node) = child {
+                            // We could be walking a partial trie, so we need to check if the node
+                            // is in the changed nodes, if it is process it
+                            if nodes.contains_key(&hash_node.hash) {
+                                stack.push(hash_node.hash);
+                                paths_to_parent.insert(
+                                    hash_node.hash,
+                                    PathToParent::new(node_key, vec![i as u8]),
+                                );
+                            }
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+
```

Comment:
Regarding previous discussion, I would add a check here:
```rust
// Ensure we discovered all nodes (root is not present in the paths_to_parent)
ensure!(nodes.len() == paths_to_parent.len() + 1, "...");
```

Reply:
This is so nice good catch